models/classes/CrudResultsService.php

```diff
@@ -55,47 +59,52 @@ public function get($uri, $groupBy = self::GROUP_BY_DELIVERY)
+
+    /**
+     * @param taoResultServer_models_classes_ReadableResultStorage $resultStorage
+     * @param $resultIdentifier
+     * @param int $groupBy
+     * @param bool $lastResult
+     * @return array
+     * @throws common_exception_Error
+     */
+    public function format(taoResultServer_models_classes_ReadableResultStorage $resultStorage, $resultIdentifier, $groupBy = self::GROUP_BY_DELIVERY, $lastResult = false)
     {
         $returnData = [];
-        
+
         if ($groupBy === self::GROUP_BY_DELIVERY || $groupBy === self::GROUP_BY_ITEM) {
             $calls = $resultStorage->getRelatedItemCallIds($resultIdentifier);
         } else {
             $calls = $resultStorage->getRelatedTestCallIds($resultIdentifier);
+            $lastResult = false;
         }
 
         foreach ($calls as $callId) {
             $results = $resultStorage->getVariables($callId);
-            $resource = [];
+            $lastData = [];
             foreach ($results as $result) {
                 $result = array_pop($result);
                 if (isset($result->variable)) {
-                    $resource['value'] = $result->variable->getValue();
-                    $resource['identifier'] = $result->variable->getIdentifier();
-                    if ($result->variable instanceof \taoResultServer_models_classes_ResponseVariable) {
-                        $type = "http://www.tao.lu/Ontologies/TAOResult.rdf#ResponseVariable";
-                    } else {
-                        $type = "http://www.tao.lu/Ontologies/TAOResult.rdf#OutcomeVariable";
+                    $resource = $this->formResource($result->variable);
+                    if (!$lastResult) {
+                        $lastData[] = $resource;
+                    } elseif ((empty($lastData[$resource['identifier']])
+                        || preg_replace('/0(\.\d*)\s(\d*)/', '$2$1', $resource['epoch'])
+                        > preg_replace('/0(\.\d*)\s(\d*)/', '$2$1', $lastData[$resource['identifier']]['epoch']))) {
+                        $lastData[$resource['identifier']] = $resource;
                     }
-                    $resource['type'] = new \core_kernel_classes_Class($type);
-                    $resource['epoch'] = $result->variable->getEpoch();
-                    $resource['cardinality'] = $result->variable->getCardinality();
-                    $resource['basetype'] = $result->variable->getBaseType();
-                }
-
-                if ($groupBy === self::GROUP_BY_DELIVERY) {
-                    $returnData[$resultIdentifier][] = $resource;
-                } else {
-                    $returnData[$callId][] = $resource;
                 }
             }
+            if ($groupBy === self::GROUP_BY_DELIVERY) {
+                $returnData[$resultIdentifier] = array_merge($returnData[$resultIdentifier] ?? [], array_values($lastData));
```

Comment:
It is rather complex. I have no other ideas how to rewrite this part of code without refactoring whole functionality.

Reply:
```
            if ($groupBy === self::GROUP_BY_DELIVERY) {
                $returnData[$resultIdentifier][] = array_values($lastData);
            } else {
                $returnData[$callId][] = array_values($lastData);
            }
        }
        return array_merge(...$returnData);
```

Reply:
```php
$groupIndex = $groupBy === self::GROUP_BY_DELIVERY ? $resultIdentifier : $callId;
$returnData[$groupIndex][] = array_values($lastData);
```

Reply:
`return array_merge(...$returnData);` will lost keys. 
It will looks like:
`foreach ($returnData as $key=>$data){
          $newReturn[$key] = array_merge(...$data);
}`

Reply:
yes. indeed. but it is still better